# Actividad última curvas elípticas

Importación de nuestras propios códigos para ser utilizados en el programa

In [1]:
import sys
import random
sys.path.append('../')

import core.mod_math as mod
import core.elliptic_curves as ec
import core.cripto as cr

### Problema 1

Consideremos la curva elíptica E definida sobre $F_{13}$ de ecuación

$$y^2 = x^3 + 10x + 6$$

**a)** Encontrar todos los puntos de la curva elíptica  E (14 en total contando el punto del infinito) y comprobar que el punto P(1, 2) tiene orden 14.

Encontramos todos los puntos de la curva elíptica

In [2]:
# Utilizamos la función que obtiene todos los puntos
# Esta funcion devuelve la tabla con la que se realizan los cálculos
# Además de los puntos y el número total de puntos
table, points, n_points = ec.get_all_ec_points(a = 10, b = 6, p = 13)
print('Tabla de procesamiento de datos')
table

Tabla de procesamiento de datos


x^3 + Ax + B   y1, y2
x                        
0              6       []
1              4  [2, 11]
2              8       []
3             11       []
4              6       []
5             12   [5, 8]
6              9  [3, 10]
7              3   [4, 9]
8              0      [0]
9              6       []
10             1  [1, 12]
11             4  [2, 11]
12             8       []

In [3]:
print('Tabla con los puntos finales en x y y')
points

Tabla con los puntos finales en x y y


x     y
   1.0   2.0
   1.0  11.0
   5.0   5.0
   5.0   8.0
   6.0   3.0
   6.0  10.0
   7.0   4.0
   7.0   9.0
   8.0   0.0
  10.0   1.0
  10.0  12.0
  11.0   2.0
  11.0  11.0
   inf   inf

In [4]:
print(f'El número de puntos de la curva elíptica es {n_points}')

El número de puntos de la curva elíptica es 14


Comprobamos que el punto P(1, 2) tiene orden 14

In [7]:
print(f'El orden el punto es: {ec.order_point(x = 1, y = 2, a = 10, p = 13)}')

El orden el punto es: 14


Como se observa, el punto P(1, 2) tiene en efecto orden 14

**b)** Si queremos cifrar y descifrar mensajes con el esquema ElGamal elíptico con parámetros (p, a, b, P, n) = (13, 10, 6, (5, 5), 7)  y hemos elegido como clave privada d=4 ¿cuál es nuestra clave pública?

La clave pública se encuentra utilizando el punto generador $P(5, 5)$ y multiplicándolo por la clave privada d = 4

In [8]:
print(f'La clave pública es: {ec.mult_finite(n = 4, x = 5, y = 5, a = 10, p = 13)}')

La clave públic es: (11, 2)


**c)** Cifrar el mensaje m = 10 con la clave pública del apartado anterior

In [14]:
# Seleccionamos un valor aleatorio para la clave efimera K entre 1 y n-1, en donde n = 7
k = random.randint(1, 6)
print(f'La llave efimera es: {k}')

# Generamos la firma dígital r, s
print(f'La firma dígital es {cr.ECDSA(x = 5, y = 5, a = 10, q = 13, d = 4, k = k, n = 7, m = 10)}')


La llave efimera es: 1
La firma dígital es (5, 2)


### Problema 2

Consideremos la curva elíptica $E$ definida sobre $F_{347}$  de ecuación:

$$y^2 = x^3 + 333x + 2$$

y el punto $P(110, 136)$

In [21]:
ec.order_point(2, 26, 333, 347)

358

**a)** Sabiendo que el cardinal de la curva es 358. ¿podemos decir que la curva es criptográficamente útil? ¿Cuál es el orden del punto P? ¿Entre qué posibles valores se puede escoger la clave privada?

Orden el punto P(110, 136) sobre la curva elíptica

In [23]:
print(f'EL orden del punto P(110, 136) es: {ec.order_point(110, 136, 333, 347)}')

EL orden del punto P(110, 136) es: 179


Para saber si la curva es criptográficamente útil, es necesario más información que solo el cardinal de la curva. Con sólo el cardinal podemos saber que no es un cardinal pequeño y que por tanto tiene muy buenas chances de ser criptográficamente útil, pero si es realmente útil o no dependerá también del orden del punto que se escoja como generador. Para este caso, el punto generador es P(110, 136). En este caso los posibles valores entre los cuales podemos escoger una clave privada están entre el 1 y el orden de ese punto menos 1, es decir, entre [1, 178] ¿Por qué? Recordemos que en criptografía de curvas elípticas es necesario en algún punto de los cálculos multiplicar el punto generador por la clave privada. La clave privada tiene que ser forzosamente menor al orden el punto ya que si es mayor, la multiplicación llegará en un punto a la identidad, lo que no solo imposibilita los cálculos computacionalmente, sino que también ya no tiene caso seguir sumando, por que solo se obtendrán puntos que ya se habían obtenido por la naturaleza cíclica de la suma en curvas elípticas discretas.

La seguridad de una curva elíptica esta dadá por que tan díficil es encontrar la llave privada de la información pública que tenemos, es decir, resolver el problema de DH aplicado a curvas elípticas. En este caso, sabemos que la clave pública es algo como esto: Q = sP, donde P es el punto generador, Q es la clave pública y s es la clave privada. Si quisieramos obtener s, entonces tendríamos que probar para cada punto todas las multiplicaciones posibles. Para esta curva elíptica con 358 puntos habría que probar varias multiplicaciones, teniendo que multiplicar en algunos casos hasta 179 que es el orden de este punto. Claramente no todos los puntos de la curva elíptica tienen el mismo orden y algunos puntos pueden tener ordenes pequeños o grandes. Si supusieramos que todos los puntos tienen el mismo orden de 179, entonces serían necesarias 358*179 operaciones para resolver el problema de DH en curvas elípticas. Esto está en el orden de 100,000 operaciones. Esto significa que la curva no es útil criptográficamente para entornos donde existan computadoras decentes, ya que estás podrían resolver el problema del DH en un tiempo muy reducido. 

**b)** Si tu clave privada es d = 101 y  un amigo te ha enviado el mensaje cifrado:

$$(C_1 = (232, 278), C_2 = (135, 214) )$$

¿cuál es el mensaje original?

In [7]:
print(f'El mensaje original es: {cr.decipher_ec_elgamal((232, 278), (135, 214), 101, 333, 347)}')

El mensaje original es: (74, 87)
